consensus/hotstuff/signature/weighted_signature_aggregator.go

```diff
@@ -1,39 +1,155 @@
+	for _, id := range signerIDs {
+		publicKeys = append(publicKeys, idToKey[id])
+	}
+
+	// build a low level aggregator
+	agg, err := signature.NewSignatureAggregatorSameMessage(message, []byte(dsTag), n, publicKeys)
+	if err != nil {
+		return nil, fmt.Errorf("new signature aggregator failed: %w", err)
+	}
+
+	// build the weighted aggregator
+	WeightedAgg := &WeightedSignatureAggregator{
+		SignatureAggregatorSameMessage: agg,
+		signerIDs:                      signerIDs,
+		weights:                        weights,
+	}
+
+	// build the idToIndex map
+	for i, id := range signerIDs {
+		WeightedAgg.idToIndex[id] = i
+	}
+	return WeightedAgg, nil
+}
 
-// Verify returns true if and only if the signature is valid.
-// It expects that correct type of signature is passed. Only SigTypeStaking is supported
+// Verify verifies the signature under the stored public and message.
 func (s *WeightedSignatureAggregator) Verify(signerID flow.Identifier, sig crypto.Signature) (bool, error) {
-	panic("to be implemented")
+	index, ok := s.idToIndex[signerID]
+	if !ok {
+		return false, fmt.Errorf("couldn't find signerID %s in the index map", signerID)
+	}
+	return s.SignatureAggregatorSameMessage.Verify(index, sig)
 }
 
-// TrustedAdd adds an already verified signature, and look up the weight for the given signer,
-// and add it to the total weight, and returns the total weight that have been collected.
-// return (1000, nil) means the signature has been added, and 1000 weight has been collected in total.
-// return (1000, nil) means the signature is a duplication and 1000 weight has been collected in total.
-func (s *WeightedSignatureAggregator) TrustedAdd(signerID flow.Identifier, weight uint64, sig crypto.Signature) (totalWeight uint64, exception error) {
-	panic("to be implemented")
+// TrustedAdd adds a signature to the internal set of signatures.
+//
+// It adds the signer's weight to the total collected weight and returns the total weight regardless
+// of the returned error.
+// The function errors if a signature from the signerID was already collected.
+// The function is thread-safe
+func (s *WeightedSignatureAggregator) TrustedAdd(signerID flow.Identifier, sig crypto.Signature) (uint64, error) {
+	// get the total weight safely
+	collectedWeight := s.TotalWeight()
+
+	// get the index
+	index, ok := s.idToIndex[signerID]
+	if !ok {
+		return collectedWeight, fmt.Errorf("couldn't find signerID %s in the index map", signerID)
+	}
+	// get the weight
+	weight, ok := s.weights[signerID]
+	if !ok {
+		return collectedWeight, fmt.Errorf("couldn't find signerID %s in the weight map", signerID)
+	}
+
+	// atomically update the signatures pool and the total weight
+	s.lock.Lock()
+	collectedWeight = s.totalWeight
+
+	// This is a sanity check because the upper layer should have already checked for double-voters.
+	_, ok = s.collectedIDs[signerID]
+	if ok {
+		s.lock.Unlock()
+		return collectedWeight, fmt.Errorf("SigneID %s was already added", signerID)
+	}
```

Comment:
We don't really need to handle duplicates in the low level crypto, as we can assume the caller is responsible for duplicates (the function is `Trusted` anyway). The new signature could replace the previously added signature from the same signer. 
At the same time, the check is cheap and we could do it since the crypto component supports signatures of multiplicity one only.